relayer/chain/relaychain/connection.go

```diff
@@ -104,102 +104,79 @@ func (co *Connection) GetMMRLeafForBlock(
-	// here as a hack.
-	actualBaseParaHeadsStorageKey := baseParaHeadsStorageKey[:32]
-	co.log.WithField("actualBaseParaHeadsStorageKey", actualBaseParaHeadsStorageKey.Hex()).Info("actualBaseParaHeadsStorageKey")
+	keyPrefix := types.CreateStorageKeyPrefix("Paras", "Heads")
 
-	keysResponse, err := co.GetAPI().RPC.State.GetKeys(actualBaseParaHeadsStorageKey, blockHash)
+	keys, err := co.GetAPI().RPC.State.GetKeys(keyPrefix, blockHash)
 	if err != nil {
 		co.log.WithError(err).Error("Failed to get all parachain keys")
-		return nil, 0, types.Header{}, err
+		return nil, err
 	}
 
-	headersResponse, err := co.GetAPI().RPC.State.QueryStorage(keysResponse, blockHash, blockHash)
+	changeSets, err := co.GetAPI().RPC.State.QueryStorageAt(keys, blockHash)
 	if err != nil {
 		co.log.WithError(err).Error("Failed to get all parachain headers")
-		return nil, 0, types.Header{}, err
+		return nil, err
 	}
 
-	co.log.Info("Got all parachain headers")
-	var headers []types.Bytes
-	var ownParachainHeaderPos int
-	for _, headerResponse := range headersResponse {
-		for index, change := range headerResponse.Changes {
-
-			// TODO fix this manual slice with a proper type decode. only the last few bytes are for the ParaId,
-			// not sure what the early ones are for.
-			key := change.StorageKey[40:]
-			var parachainID types.U32
-			if err := types.DecodeFromBytes(key, &parachainID); err != nil {
-				co.log.WithError(err).Error("Failed to decode parachain ID")
-				return nil, 0, types.Header{}, err
+	heads := make(map[uint32]Head)
+
+	for _, changeSet := range changeSets {
+		for index, change := range changeSet.Changes {
+			if change.StorageData.IsNone() {
+				continue
 			}
 
-			var headerBytes types.Bytes
-			if err := types.DecodeFromBytes(change.StorageData, &headerBytes); err != nil {
-				co.log.WithError(err).Error("Failed to decode MMREncodableOpaqueLeaf")
-				return nil, 0, types.Header{}, err
+			var paraID uint32
+
+			if err := types.DecodeFromBytes(change.StorageKey[40:], &paraID); err != nil {
+				co.log.WithError(err).Error("Failed to decode parachain ID")
+				return nil, err
 			}
-			headers = append(headers, headerBytes)
 
-			if parachainID == types.U32(ownParachainId) {
-				ownParachainHeaderPos = index
+			_, headDataWrapped := change.StorageData.Unwrap()
+
+			var headData types.Bytes
+			if err := types.DecodeFromBytes(headDataWrapped, &headData); err != nil {
+				co.log.WithError(err).Error("Failed to decode HeadData wrapper")
+				return nil, err
 			}
 
+			heads[paraID] = Head{
+				LeafIndex: index,
+				ParaID:    paraID,
+				Data:      headData,
+			}
 		}
 	}
-	var ownParachainHeader types.Header
-	if err := types.DecodeFromBytes(headers[ownParachainHeaderPos], &ownParachainHeader); err != nil {
-		co.log.WithError(err).Error("Failed to decode Header")
-		return nil, 0, types.Header{}, err
-	}
 
-	co.log.WithField("parachainId", ownParachainId).Info("Decoding header for own parachain")
```

Comment:
please keep all this excessive logging in at least until we complete e2e verification. it's likely going to all be needed for debugging/completion purposes once we activate the verification in the smart contracts, as i don't expect it to just work.

also, it seems fine to even just keep the excessive logging in indefinitely?

Reply:
Will add it back :+1:

Reply:
Yeah, excessive logging is good. Added a bunch of extra logging in the latest upload.